feat: Highlight gitignored files (and fix g:nvim_tree_gitignore)

sindrets opened the pull request:
Let me first say: sorry for the somewhat large PR. If it makes it easier to review, I can split this into at least two PRs.

This PR does ~~three~~ *two* things, all related to ignored files:
- Enables highlighting files/dirs that are ignored by git.
- Fixes ignoring gitignored files/dirs when `g:nvim_tree_gitignore` is enabled.
- ~~Changes the implementation of ignored files.~~

### Highlighting

Highlight nodes in the file tree that are ignored by git when `g:nvim_tree_git_hl` is enabled.

![2021-04-03-174542_maim](https://user-images.githubusercontent.com/2786478/113483583-7e555100-94a4-11eb-856a-ebcd22d0e240.png)

### Fixed ignoring gitignored files (`g:nvim_tree_gitignore`)

The current implementation only ignores gitignored files that reside directly in the root of the file tree, and it does not ignore newly added files/dirs that are gitignored. This implementation updates the list of gitignored files when changes are made to either `.gitignore` or `.git/info/exclude`, or new nodes are added to the file tree.

### ~~Changed implementation of ignored files~~

~~Currently ignored nodes don't always keep their position in the file tree when toggled. This implementation doesn't skip creation of ignored nodes. Instead it marks them as ignored, and makes the renderer simply skip rendering ignored nodes. This has several advantages: nodes always retain their position in the tree. No need to dispose of / create nodes when toggling ignored nodes or dotfiles.~~

~~And I provide this utility method for filtering out ignored nodes from a list of entries:~~

~~https://github.com/sindrets/nvim-tree.lua/blob/a511ea08d8e5b59c44a7c2f4502b8a6913ae6009/lua/nvim-tree/lib.lua#L64-L81~~

~~The reason I added an optional parameter  - `separator` - to the path functions in `utils.lua` is that i.e. git always return paths with forward-slashes as a separator, regardless of platform. If the parameter is not provided it falls back to the platform default.~~

stsewd commented on lua/nvim-tree/git.lua:
```diff
@@ -30,8 +32,16 @@ local function update_root_status(root)
+function M.get_gitignored(git_root)
+  local cmd = "git ls-files --others --ignored --exclude-standard --directory"
+  if git_root then
+    cmd = "cd '" .. git_root .. "' && " .. cmd
```
You can pass the cwd as `-C {cwd}` in the git command, if it's empty it defaults to the cwd.

sindrets replied:
Yes, that's better. Good catch.

kyazdani42 commented on lua/nvim-tree/git.lua:
```diff
@@ -87,9 +115,18 @@ function M.update_status(entries, cwd)
+    if parent_node.git_status == "ignored" or M.should_gitignore(node.absolute_path) then
+      node.git_status = "ignored"
+      num_ignored = num_ignored + 1
+      goto continue
```
i believe we should avoid goto because it might not match all users lua version (i remember having this issue before and i had to remove the gotos).

sindrets replied:
Gotcha. I wasn't aware that it had limited support. Thanks.

kyazdani42 commented on lua/nvim-tree/git.lua:
```diff
@@ -110,6 +147,81 @@ function M.update_status(entries, cwd)
+
+      for _, s in ipairs(M.get_gitignored(git_root)) do
+        s = utils.path_remove_trailing(s, "/")
+        paths[utils.path_join({git_root, s}, "/")] = true
```
is git also adding forward slashes on windows ? if not, you should use the `utils.path_separator` to do that.

kyazdani42 reviewed (changes requested):
thanks for the PR. I would like it if you'd split this in two, so it's easier for me to review it properly :)

sindrets commented:
@kyazdani42 I have now removed the changed ignore implementation, so this PR is now just about highlighting gitignored files. And in implementing that, I also fixed the ignoring of gitignored files, as it was a relatively small change after I had added caching of gitignored files.

Regarding the changed ignore implementation: I moved those changes to a different branch, but while it does solve the problem I described, I'm not sure it's something you would want. It works just fine, but it adds some complexity in that you have to make a conscious decision on whether to iterate on all entries, or just visible entries. So let me know if you're interested in implementing it, but if not I think I'll just leave it.

kyazdani42 commented:
hi again, thanks for your contribution. I'm waiting to resolve #172 before merging this (but i think the culprit is not nvim-tree), thanks for waiting :)

kyazdani42 commented on lua/nvim-tree/populate.lua:
```diff
@@ -73,6 +73,7 @@ local function link_new(cwd, name)
     absolute_path = absolute_path,
     link_to = link_to,
     open = open,
+    group_next = nil,   -- If node is grouped, this points to the next child dir/link node
```
defining to nil is not required though, if trying to access group_next, it will be nil by default when not declared

kyazdani42 reviewed (approved):
this PR works very well though, i'm merging this :)

kyazdani42 commented:
i don't know if it would add much benefit to add the `changed ignore implementation`. anyway, good job :)
